addon/components/frost-modal-binding.js

```diff
@@ -81,13 +143,40 @@ const FrostModalBinding = Component.extend(PropTypesMixin, {
       this.onClose()
     },
 
+    /**
+     * Handle confirmation in a promise-aware way.
+     * If onConfirm returns a promise, closeOnConfirm will wait for it to resolve successfully before closing the modal.
+     * @returns {Object} the return value of onConfirm.
+     * @private
+     */
     _onConfirm () {
       const onConfirm = this.get('onConfirm')
+      let confirmed
       if (onConfirm) {
-        onConfirm()
+        confirmed = onConfirm()
       }
+
       if (this.get('closeOnConfirm') === true) {
+        if (isThenable(confirmed)) {
+          // We handle these separately because someone might want to
+          // have the confirm button disabled, but not call onClose automatically.
+          if (this.get('disableConfirmUntilOnConfirmResolves')) {
+            this.set('forceDisabledConfirm', true)
+            confirmed = confirmed.then(() => {
+              this.set('forceDisabledConfirm', false)
+            })
+          }
+
+          confirmed = confirmed.then(() => {
```

Comment:
Shouldn't this be in an `else`?

Reply:
No.  The `else` for the prior `if` is to do nothing regarding the disabled state (and would therefore be empty).  

This is merely the action we want to happen within the `if closeOnConfirm ...` branch, namely call onClose in a promise-aware way.

Reply:
```
          confirmed = confirmed.then(() => {
            this.onClose()
          })
```

is always going to run with this block of code.  The `if` statement before it

```
          if (this.get('disableConfirmUntilOnConfirmResolves')) {
            this.set('forceDisabledConfirm', true)
            confirmed = confirmed.then(() => {
              this.set('forceDisabledConfirm', false)
            })
          }
```

may optionally set `confirmed` to something as well.

So if the first `if` statement is executed it doesn't matter what `confirmed` was set to because the code setting `confirmed` after the `if` statement is always going to run.

Or am I missing something?

Reply:
Because they're in the same block, you can re-assign promises.  It's re-using the `confirmed` variable.

I previously did
```js
let thenable = confirmed

if (....) {
  thenable = confirmed.then( /* fix button */ )
}
thenable = confirmed.then( /* invoke onClose */ )
return thenable
```
but then realized that `thenable` was completely redundant, because one can do the same reassignment with `confirmed` anyway.

The end result is returning a promise that is basically
```js
confirmed
  .then(() => {...})  // invoke onClose
  .finally(() => { ... })  // fix button state
```
(or should be once I fix the bug I commented about in a sister comment :))